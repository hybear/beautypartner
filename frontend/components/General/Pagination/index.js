import React from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import PaginationStyles from './styles';
import { perPage } from '../../../config';
import Link from 'next/link';

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const Pagination = props => (
  <Query query={PAGINATION_QUERY}>
    {({ data, loading, error }) => {
      if (loading) return null;
      const count = data.itemsConnection.aggregate.count;
      const pages = Math.ceil(count / perPage);
      const page = props.page;
      return (
        <PaginationStyles data-test="pagination">
          <Link
            // prefetch
            href={{ query: { page: page - 1 } }}
          >
            <a className="prev" aria-disabled={page <= 1}>
              {' '}
              Prev{' '}
            </a>
          </Link>
          <p>
            Page {page} of <span data-test="totalPages">{pages}</span>
          </p>
          <p>{count} Products Total</p>
          <Link
            // prefetch
            href={{ query: { page: page + 1 } }}
          >
            <a className="next" aria-disabled={page >= pages}>
              Next
            </a>
          </Link>
        </PaginationStyles>
      );
    }}
  </Query>
);

export default Pagination;
export { PAGINATION_QUERY };
