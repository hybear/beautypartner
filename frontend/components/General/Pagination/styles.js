import styled from 'styled-components';

const PaginationStyles = styled.div`
  text-align: center;
  display: grid;
  grid-template-columns: repeat(4, auto);
  align-items: stretch;
  justify-content: center;
  align-content: center;
  margin: 2rem auto;
  border: 1px solid ${({ theme }) => theme.color.lightGrey};
  border-radius: 10px;
  width: fit-content;
  ${({ theme }) => theme.media.tablet`
      width: 100%;
  `}
  & > * {
    margin: 0;
    padding: 15px 30px;
    border-right: 1px solid ${({ theme }) => theme.color.lightGrey};
    &:last-child {
      border-right: 0;
    }
    ${({ theme }) => theme.media.tablet`
      padding: 15px 10px;
    `}
  }
  a[aria-disabled='true'] {
    color: grey;
    pointer-events: none;
  }
`;

export default PaginationStyles;
