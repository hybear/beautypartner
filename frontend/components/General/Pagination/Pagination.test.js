import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';
import Pagination, { PAGINATION_QUERY } from './index'
import Router from 'next/router';

Router.router = {
    push() {},
}

function makeMocksFor(length){
    return  [
        {
            request: { query: PAGINATION_QUERY },
            result: {
                data: {
                    itemsConnection: {
                        __typename: 'aggregate',
                        aggregate: {
                            __typename: 'count',
                            count: length
                        }
                    }
                }
            }
        }
    ]
}

describe('<Pagination/>', () => {
    it('displays a loading message', () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(1)}>
                <Pagination page={1} />
            </MockedProvider>
        )
        expect(wrapper.text()).toContain('Loading...');
        // expect(pagination).toMatchSnapshot();
        // console.log(wrapper.debug());
    })
    
    it('renders pagination for 20 items', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(20)}>
                <Pagination page={1} />
            </MockedProvider>
        )
        await wait();
        wrapper.update();
        
        const totalPages = wrapper.find('[data-test="totalPages"]');
        expect(totalPages.text()).toEqual('5')
        expect(totalPages).toMatchSnapshot();
    });

    it('disables prev button on first page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(20)}>
                <Pagination page={1} />
            </MockedProvider>
        )
        await wait();
        wrapper.update();

        expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(true);
        expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
    })

    it('disables next button on last page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(20)}>
                <Pagination page={5} />
            </MockedProvider>
        )
        await wait();
        wrapper.update();

        expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
        expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(true);
    })

    it('enables all button on middle page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(20)}>
                <Pagination page={3} />
            </MockedProvider>
        )
        await wait();
        wrapper.update();

        expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
        expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
    })
})