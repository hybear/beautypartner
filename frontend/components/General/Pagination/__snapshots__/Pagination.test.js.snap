// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> renders pagination for 20 items 1`] = `
<span
  data-test="totalPages"
>
  5
</span>
`;
